More CoreFx porting work

jaredpar opened the pull request:
This is more progress on moving Microsoft.CodeAnalysis.Desktop to be a CoreFx compatible assembly.

jaredpar commented:
@agocke, @jasonmalinowski, @VSadov, @tmat, @theoy

VSadov commented:
LGTM

tmat commented:
:+1:

pharring commented on src/Compilers/Core/Desktop/EncodedStringText.cs:
```diff
@@ -37,10 +37,11 @@ internal static class EncodedStringText
         {
             Debug.Assert(stream != null);
             Debug.Assert(stream.CanRead && stream.CanSeek);
+            Debug.Assert(defaultEncoding != null);
```
I don't think this is right. It neuters the "detectEncoding" logic below.

pharring replied:
I presume you had to touch this because Encoding.Default is not available on CoreCLR.

tmat replied:
Paul is correct. I missed that. We might need additional bool detectEncoing parameter

jaredpar replied:
Hmm, yes you're correct.  I will update that.

jaredpar commented:
Updated in response to the `Encoding.Default` portion of the change.  I switched it to use a more explicit last ditch fallback of `Encoding.ASCII`.

Note: The use of ASCII was decided upon during an informal conversation with @pharring.  After working on this change though I'm wondering if extended ASCII (ISO 8859-1, Windows 1252, etc...) would be a better choice here.

pharring commented:
If ISO 8859-1 (Latin1) is available on CoreCLR, then I agree.

agocke commented:
Yes, codepage 1252 is better if it's available.

agocke commented on src/Compilers/Core/Desktop/EncodedStringText.cs:
```diff
@@ -51,13 +51,13 @@ internal static SourceText Create(Stream stream, Encoding defaultEncoding = null
             try
             {
-                return Decode(stream, defaultEncoding ?? Encoding.Default, checksumAlgorithm, throwIfBinaryDetected: detectEncoding);
+                return Decode(stream, defaultEncoding ?? Encoding.ASCII, checksumAlgorithm, throwIfBinaryDetected: detectEncoding);
```
`Encoding.GetEncoding(codepage: 1252)`?

agocke commented on src/Compilers/Core/Desktop/EncodedStringText.cs:
```diff
@@ -23,7 +23,7 @@ internal static class EncodedStringText
         /// 1. It attempts to minimize allocations by trying to read the stream into a byte array.
         /// 2. If <paramref name="defaultEncoding"/> is null, it will first try UTF8 and, if that fails, it will
-        ///    try <see cref="Encoding.Default"/>.
+        ///    try <see cref="Encoding.ASCII"/>.
```
I don't see how we could possibly succeed at getting ASCII but failing to get UTF8, since they're backwards compatible.

jaredpar commented:
The code page 1252 does appear to be available so I will use that.
